### PR TITLE
fix a few corner cases and revert last batch for `SpanBatch` when error happens

### PR DESCRIPTION
**Pull request by zhiqiangxu**
According to [this code](https://github.com/ethereum-optimism/op-geth/blob/optimism/miner/worker.go#L968-L971), `parent.Timestamp==current.Timestamp` is also invalid.

This PR also fixes `checkFull` so that it only marks `ErrCompressorFull` when `co.compressed.Len() > co.target` instead of `co.compressed.Len() >= co.target`, because when `co.compressed.Len() == co.target`, we actually want to include the current block into the channel instead of trying to [revert](https://github.com/ethereum-optimism/optimism/blob/d2263b19958f7f1e738ff79dae7ebb240cca1375/op-node/rollup/derive/span_channel_out.go#L166) to the compressed previous data.

This PR also added a call to `revertLastBatch` when error happened downstream.

### CODE REVIEW

File: op-node/rollup/derive/span_batch.go [hunk 1]
```diff
@@ -505,7 +505,7 @@ func (b *SpanBatch) peek(n int) *SpanBatchElement { return b.Batches[len(b.Batch
 // updates l1OriginCheck or parentCheck if needed.
 func (b *SpanBatch) AppendSingularBatch(singularBatch *SingularBatch, seqNum uint64) error {
 	// if this new element is not ordered with respect to the last element, panic
-	if len(b.Batches) > 0 && b.peek(0).Timestamp > singularBatch.Timestamp {
+	if len(b.Batches) > 0 && b.peek(0).Timestamp >= singularBatch.Timestamp {
```
Comment (trianglesphere) [inline op-node/rollup/derive/span_batch.go:512]: I think this case would only happen if we added a batch twice which I think we may actually want to prevent cc @axelKingsley

**Comment by trianglesphere** [comment]
@zhiqiangxu op-batcher tests are failing with this change

**Comment by zhiqiangxu** [comment]
I added a `SkipOptimize` which is only set to true for test, as I met a case where the size before compressing(`969`) is smaller than the size after compressing(`985`). In this situation there's no way to choose a proper `MaxFrameSize` to make `co.target` smaller than `969` but bigger than `985`.

I also found a couple other issues, described at the beginning of the post.

File: op-node/rollup/derive/span_batch.go (hunk 1 above)
Comment (axelKingsley) [reply op-node/rollup/derive/span_batch.go:512]: I agree, adding the same batch twice would be incorrect and is a valid panic condition. 

I still need to read over the rest of this PR, specifically the revert code. It's a good feature to add but it can be tricky to ensure the bitmaps get reverted correctly.

File: op-batcher/batcher/channel_builder_test.go [hunk 2]
```diff
@@ -552,11 +563,15 @@ func TestChannelBuilder_FullShadowCompressor(t *testing.T) {
 	require.NoError(err)
 
 	rng := rand.New(rand.NewSource(420))
-	a := dtest.RandomL2BlockWithChainId(rng, 1, defaultTestRollupConfig.L2ChainID)
+	blockTime := time.Now()
+	a := dtest.RandomL2BlockWithChainIdAndTime(rng, 1, defaultTestRollupConfig.L2ChainID, blockTime.Add(time.Duration(1)*time.Second))
 	_, err = cb.AddBlock(a)
 	require.NoError(err)
+	require.Nil(cb.fullErr)
```
Comment (axelKingsley) [inline op-batcher/batcher/channel_builder_test.go]: As `fullErr` is an error, you can use `require.NoError` like above.

File: op-node/rollup/derive/params.go [hunk 3]
```diff
@@ -38,8 +38,8 @@ const MaxChannelBankSize = 100_000_000
 // a channel. This limit is set when decoding the RLP.
 const MaxRLPBytesPerChannel = 10_000_000
 
-// DuplicateErr is returned when a newly read frame is already known
-var DuplicateErr = errors.New("duplicate frame")
+// ErrDuplicate is returned when a newly read frame is already known
+var ErrDuplicate = errors.New("duplicate frame")
```
Comment (axelKingsley) [inline op-node/rollup/derive/params.go:42]: is there a specific reason this symbol is being renamed? I didn't notice it in the PR description.

File: op-node/rollup/derive/span_batch.go [hunk 4]
```diff
@@ -505,15 +509,20 @@ func (b *SpanBatch) peek(n int) *SpanBatchElement { return b.Batches[len(b.Batch
...
 	// always append the new batch and set the L1 origin check
 	b.Batches = append(b.Batches, singularBatchToElement(singularBatch))
 
 	// always update the L1 origin check
+	copy(b.lastL1OriginCheck[:], b.L1OriginCheck[:])
 	copy(b.L1OriginCheck[:], singularBatch.EpochHash.Bytes()[:20])
+
+	// it's easy to tend to only update lastParentCheck when ParentCheck is updated,
+	// but it's wrong because if done that way, lastParentCheck will always be empty.
+	copy(b.lastParentCheck[:], b.ParentCheck[:])
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch.go]: I would rephrase this comment or remove it altogether. Explaining that something is "easy but wrong" is unnecessary. I am okay with an explanation for why this must always be copied, but I actually don't understand what you mean about it always being empty.

File: op-node/rollup/derive/span_batch.go [hunk 5]
```diff
@@ -544,6 +553,33 @@ func (b *SpanBatch) AppendSingularBatch(singularBatch *SingularBatch, seqNum uin
 	return b.sbtxs.AddTxs(newTxs, b.ChainID)
 }
 
+func (b *SpanBatch) revertLastBatch() {
+	if len(b.Batches) == 0 {
+		panic("revertLastBatch should only be called when there's at least 1 batch")
+	}
+	// revert ParentCheck
+	copy(b.ParentCheck[:], b.lastParentCheck[:])
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch.go:561]: What happens if a caller attempts to revert the last batch multiple times? Looks like `lastParentCheck` will remain, and while the batches will roll back, this value will persist.

You'd need to keep an entire slice of `ParentCheck`s in order to support reverting more than once, right?

File: op-node/rollup/derive/span_batch_test.go [hunk 6]
```diff
@@ -602,3 +602,42 @@ func TestSpanBatchTotalBlockTxCountNotOverflow(t *testing.T) {
...
+	singularBatches := RandomValidConsecutiveSingularBatches(rng, chainID)
+	// initialize empty span batch
+	spanBatch := initializedSpanBatch([]*SingularBatch{}, uint64(0), chainID)
+
+	for i := 0; i < len(singularBatches); i++ {
+		err := spanBatch.AppendSingularBatch(singularBatches[i], uint64(i))
+		require.NoError(t, err)
+		spanBatch.revertLastBatch()
+
+		spanBatch2 := initializedSpanBatch(singularBatches[:i], uint64(0), chainID)
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch_test.go:620]: Rather than appending and removing the same batch (and then adding it back at the end of the loop), why not test as follows:

* `spanBatch1` is made of all batches [0:n]
* `spanBatch2` is made of batches [0:i], for each `i` from 0->n
* for each `i`, revert `spanBatch1` until it only contains batches [0:i]
* Test equality

This way you could confirm that reverting works to any depth. In the current test, it only adds a singular batch to be reverted on each loop, which is kinda the same test every time.

File: op-node/rollup/derive/span_batch_tx.go [hunk 7]
```diff
@@ -69,7 +69,7 @@ func (tx *spanBatchTx) MarshalBinary() ([]byte, error) {
 }
 
 // setDecoded sets the inner transaction after decoding.
-func (tx *spanBatchTx) setDecoded(inner spanBatchTxData, size uint64) {
+func (tx *spanBatchTx) setDecoded(inner spanBatchTxData, _ /*size*/ uint64) {
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch_tx.go]: suggest just removing `/*size*/`, and perhaps making it a comment on the function that size is required for the interface, but unused (actually, *is* it required for the interface?)

File: op-node/rollup/derive/span_batch_txs.go [hunk 8]
```diff
@@ -468,3 +468,40 @@ func (sbtx *spanBatchTxs) AddTxs(txs [][]byte, chainID *big.Int) error {
 	sbtx.totalBlockTxCount += totalBlockTxCount
 	return nil
 }
+
+func (sbtx *spanBatchTxs) revertLastBatch(txs []hexutil.Bytes) {
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch_txs.go]: `spanBatchTxs` don't have a native concept of a "Batch", so it doesn't make sense to call this `revertLastBatch`, especially when the parameter is `txs`. Suggest to name this `revertTxs`

File: op-node/rollup/derive/span_channel_out.go [hunk 9]
```diff
@@ -152,9 +161,11 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
...
 	// this optimizes out cases where the compressor will obviously come in under the target size
-	rlpGrowth := co.activeRLP().Len() - co.lastCompressedRLPSize
-	if uint64(co.compressed.Len()+rlpGrowth) < co.target {
-		return nil
+	if !SkipOptimize {
+		rlpGrowth := co.activeRLP().Len() - co.lastCompressedRLPSize
+		if uint64(co.compressed.Len()+rlpGrowth) < co.target {
+			return nil
+		}
 	}
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go]: I am not in favor of flags which change component behavior for unit tests. There should be other ways of testing by using a different `rng`, thus making differently sized structures which won't have the issue you described?

File: op-node/rollup/derive/span_channel_out.go [hunk 10]
```diff
@@ -130,6 +133,12 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
 	if err := co.spanBatch.AppendSingularBatch(batch, seqNum); err != nil {
 		return fmt.Errorf("failed to append SingularBatch to SpanBatch: %w", err)
 	}
+	defer func() {
+		if err != nil {
+			co.spanBatch.revertLastBatch()
+		}
+	}()
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go:137]: Span Channel Out already blocks new Batches from being added if it is already full, so I'm not sure why we'd want to revert the spanBatch. I guess if there was a failure to `rlp` encode or something, this would roll back that batch, but this revert would quietly drop a batch from a channel, which the Channel Builder now needs to understand and handle.

I could see a situation where reverting makes sense -- you wouldn't want to add the same batch to the SpanBatch twice, and then have that be part of the encoded structure. But I could also see this complexity lead to batches being dropped -- if the error being returned is `FullErr`, this revert would mutate the underlying structure (which could be okay since it is already encoded and compressed, but still has a smell to it)

File: op-node/rollup/derive/span_batch_txs.go [hunk 11]
```diff
@@ -468,3 +468,40 @@ func (sbtx *spanBatchTxs) AddTxs(txs [][]byte, chainID *big.Int) error {
...
+		var tx types.Transaction
+		if err := tx.UnmarshalBinary(txs[len(txs)-1-idx] /* iterate in reverse order */); err != nil {
+			panic("impossible since already checked by AddTxs")
+		}
+
+		if tx.Type() == types.LegacyTxType {
+			// revert protectedBits and totalLegacyTxCount
+			sbtx.protectedBits.SetBit(sbtx.protectedBits, int(sbtx.totalLegacyTxCount)-1, 0)
+			sbtx.totalLegacyTxCount--
+		}
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_batch_txs.go:484]: I would really like some more concerted unit tests around this logic -- reversing the legacy tx bitmap (and all the other bitmaps) is a sensitive operation. I see we have a test for SpanBatchRevert, but since this is the more difficult workload, and could use its own tests.

I'll also think more about this logic, as I don't yet fully understand how it's working.

**Review by axelKingsley — CHANGES_REQUESTED** [review]
Some really cool ideas in this PR, thanks for putting it together! I have some comments around testing and edge-cases, but also wanted to leave a high-level comment:

Span Batch Reverting is a helpful feature, but I think the amount of complexity it brings in does not currently justify its benefit. Seems like the largest benefit is that it prevents Span Batches from being added to during an error in `AddSingularBatch`, because presumably on retry the same batch would be added *again*. To address this with less complexity, the SpanBatch itself could check,
* If the same timestamp as the last batch, and the same content (hash), just ignore it
* If the same timestamp as the last content, and *different* content, panic

This would cover cases where an error in the Span Channel Out might redrive content into the SpanBatch, and would not require `defer` behavior to mutate the SpanBatch. 

We could even make it something the Span Channel Out manages, with a new function, `func (sb SpanBatch)AlreadyHas(b Batch) bool`, so the SpanChannelOut could avoid redriving batches altogether.

Span Batches don't currently *need* the ability to revert, because the Span Channel Out keeps a double-buffer of RLP encoded Raw Span Batches. If the newly added batch exceeds the threshold, the previous RLP buffer is used. This is like a revert, but at the serialized level.

File: op-node/rollup/derive/params.go (hunk 3 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/params.go:42]: VScode is complaining: `error var DuplicateErr should have name of the form ErrFoo (ST1012).` So I slipped in this lint change.

File: op-node/rollup/derive/span_batch.go (hunk 4 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch.go]: Because `ParentCheck` will only be updated once, so if `lastParentCheck` is also updated only once, it'll be the initial state of  `ParentCheck`, which is empty. I don't have a simple sentence that comes to my mind, so I deleted this comment.

File: op-batcher/batcher/channel_builder_test.go (hunk 2 above)
Comment (zhiqiangxu) [reply op-batcher/batcher/channel_builder_test.go]: Done.

File: op-node/rollup/derive/span_batch.go (hunk 5 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch.go:561]: `revertLastBatch` can only revert the changes by the latest AppendSingularBatch call, it can not revert changes by previous AppendSingularBatch calls. 

I also added this statement as comment for `revertLastBatch`.

File: op-node/rollup/derive/span_batch_tx.go (hunk 7 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch_tx.go]: It seem unnecessary, I removed it .

File: op-node/rollup/derive/span_batch_txs.go (hunk 8 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch_txs.go]: I renamed it to `revertLastTxs` since it can only revert the `last` txs.

File: op-node/rollup/derive/span_batch_test.go (hunk 6 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch_test.go:620]: `revertLastBatch` can only revert the changes by the latest AppendSingularBatch call, it can not revert changes by previous AppendSingularBatch calls.

File: op-node/rollup/derive/span_channel_out.go (hunk 10 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_channel_out.go:137]: In ChannelBuilder, `block` is only [added](https://github.com/ethereum-optimism/optimism/blob/53afefdd2e53dac3e7597e58bb78132ccee9d2c4/op-batcher/batcher/channel_builder.go#L166) if `AddSingularBatch` doesn't return an error, so I think it's better to be consistent that if an error is returned, `co.spanBatch` should revert to its previous state.

File: op-node/rollup/derive/span_batch_txs.go (hunk 11 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_batch_txs.go:484]: It's basically the symmetric reverse of `AddTxs`, the main caveat is that the reverted spanBatch can not be compared with its previous state directly, as two big.Int(s) of the same math value can have different representations. That's why in `TestSpanBatchRevert` I compare two spanBatch after serialization.

File: op-node/rollup/derive/span_channel_out.go (hunk 9 above)
Comment (zhiqiangxu) [reply op-node/rollup/derive/span_channel_out.go]: Good point, let me try.

**Comment by zhiqiangxu** [comment]
> Some really cool ideas in this PR, thanks for putting it together! I have some comments around testing and edge-cases, but also wanted to leave a high-level comment:
> 
> Span Batch Reverting is a helpful feature, but I think the amount of complexity it brings in does not currently justify its benefit. Seems like the largest benefit is that it prevents Span Batches from being added to during an error in `AddSingularBatch`, because presumably on retry the same batch would be added _again_. To address this with less complexity, the SpanBatch itself could check,
> 
> * If the same timestamp as the last batch, and the same content (hash), just ignore it
> * If the same timestamp as the last content, and _different_ content, panic
> 
> This would cover cases where an error in the Span Channel Out might redrive content into the SpanBatch, and would not require `defer` behavior to mutate the SpanBatch.
> 
> We could even make it something the Span Channel Out manages, with a new function, `func (sb SpanBatch)AlreadyHas(b Batch) bool`, so the SpanChannelOut could avoid redriving batches altogether.
> 
> Span Batches don't currently _need_ the ability to revert, because the Span Channel Out keeps a double-buffer of RLP encoded Raw Span Batches. If the newly added batch exceeds the threshold, the previous RLP buffer is used. This is like a revert, but at the serialized level.

I think this argument is based on assumptions on what kind of error can actually happen downstream, right?

IMHO we'll achieve cleaner separation of concern if we can get rid of such assumptions.
